### ISSUE TITLE
FR - Growing latencies on copy receipts

### ISSUE THREAD

**Issue by mertcanyalhi** [fiskaltrust/middleware#242]
# Summary

Although the recent change (#192) improved the performance of the signature requests for copy receipts, we're observing an increasing trend for copy receipt latencies, which may start causing performance issues in our integrations.

We have outlets with >200K queue items and a `CNumerator` of nearly 5K. We noticed that the latencies have direct correlation with the number of queue items, which can be seen in the table below.

The table shows the correlation of copy receipt latency (red, left y-axis) and number of queue items (yellow, right y-axis) for multiple outlets (x-axis):

![image](https://github.com/fiskaltrust/middleware/assets/11897242/8fdd6926-9ab8-4dc0-9f5e-122cab7666b5)

# Impact

The impact can be seen in the sub-sections for each store. Each sub-section shows the queue size, `CNumerator`, and a latency heatmap for:

- All receipts - signature requests for all receipts
- Copy receipts - signature requests for copy receipts
- Other receipts - signature requests for non-copy receipts

TL;DR:
- Copy receipt latency has a direct correlation with the queue size.
- While the copy receipt request is being processed, it blocks the queue and impacts latency of other receipt requests.
  - This is expected due to the blocking nature of the queue; the main idea is to highlight the spikes in other receipt types are actually a result of copy receipt requests.
- Growing copy receipt latencies can be clustered in 2 groups;
  - First group that has higher latency
  - Second group that has lower latency, which is mainly a subsequent copy receipt request. Subsequent copy receipt requests that are made in a certain window is always lower. This can be the result of an index cache on the data layer

## Store A

- Queue size: 216K
- CNumerator: 4.7K

Latency heatmap for all receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/1837f014-0d30-4ec3-8089-d2c32f0eefcd)

Latency heatmap for copy receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/2025680e-e5d0-4c64-8a00-67ef7158b904)

Latency heatmap for other receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/43b3658e-bbaa-4a8f-8b0c-0d87f051a71d)

When we focus on the copy receipt latencies, it can be seen that the latency of the subsequent copy receipt requests that are made in a certain window is always lower. This can be the result of an index cache on the data layer. The table below shows copy receipt requests that are grouped in equal time windows:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/a2116c20-afa9-40d2-8ffe-0fb580ad1a21)

Findings:

- The latency heatmap for copy receipts indicates that the latencies can be clustered in levels; 4-5s and 10-15s
- The latency heatmap for other receipts indicates that the latencies are mostly in the same level (~500ms).
  - Latencies of other receipts seem to be similar (~500ms) for all receipt types. This can also be seen within table under the summary section.
  - In some cases, an increasing latency can be observed for other receipts. This is a result of the previous copy receipt request blocking the middleware while it is being processed.

## Store B

- Queue size: 194K
- CNumerator: 3.7K

Latency heatmap for all receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/7a69f252-07cc-4f87-9d5a-be41a1015263)

Latency heatmap for copy receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/635cfe5b-b1ca-4531-a833-bb341aaadea0)

Latency heatmap for other receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/276a2c9b-31e0-4360-bd40-e07547e5a912)

Findings:

- Same as store A.

## Store I

- Queue size: 42K
- CNumerator: 0.9K

Latency heatmap for all receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/89b966b7-04c0-4331-90be-71381d69d4ff)

Latency heatmap for copy receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/c32c24ad-13b3-41d6-95b6-8c57bb860ff5)

Latency heatmap for other receipts:

![image](https://github.com/fiskaltrust/middleware/assets/11897242/c1dde4c3-b755-47ce-a7dc-72e4f6b2b064)

Findings:

- Copy receipt request latencies are on the same level (~2.4s-3.1s).
- Other receipt request latencies are on the same level (~500ms) with spikes up to ~2.8s that is a side effect of the copy receipt request that is processed in the background.

# Expectations

As SaaS and PaaS providers, our expectation is to have stable latencies regardless of the queue size. Growing queue size will cause delays or even interruptions in receipt printing, and as a result, degraded customer experience.

**Comment by mertcanyalhi** [comment]
Just wanted to check if there is any update, or anything we can help with 🙏 

CC: @TSchmiedlechner @volllly

**Comment by mertcanyalhi** [comment]
Hey @TSchmiedlechner, thanks for the information and your efforts 🙏 please let me know if we can provide any further input or help you in any way

**Comment by StefanKert** [comment]
This should be fixed and rolled out .
